zyte_spider_templates/spiders/job_posting.py

```diff
@@ -0,0 +1,269 @@
+    }
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler, *args, **kwargs) -> Self:
+        spider = super().from_crawler(crawler, *args, **kwargs)
+        spider._init_input()
+        spider._init_extract_from()
+        return spider
+
+    def _init_input(self):
+        urls_file = self.args.urls_file
+        if urls_file:
+            response = requests.get(urls_file)
+            urls = load_url_list(response.text)
+            self.logger.info(f"Loaded {len(urls)} initial URLs from {urls_file}.")
+            self.start_urls = urls
+        elif self.args.urls:
+            self.start_urls = self.args.urls
+        else:
+            self.start_urls = [self.args.url]
+        self.allowed_domains = list(set(get_domain(url) for url in self.start_urls))
+
+    def _init_extract_from(self):
+        if self.args.extract_from is not None:
+            self.settings.set(
+                "ZYTE_API_PROVIDER_PARAMS",
+                {
+                    "jobPostingOptions": {"extractFrom": self.args.extract_from},
+                    "jobPostingNavigationOptions": {
+                        "extractFrom": self.args.extract_from
+                    },
+                    **self.settings.get("ZYTE_API_PROVIDER_PARAMS", {}),
+                },
+                priority=ARG_SETTING_PRIORITY,
+            )
+
+    def get_start_request(self, url):
+        callback = (
+            self.parse_job_posting
+            if self.args.crawl_strategy == JobPostingCrawlStrategy.direct_item
+            else self.parse_navigation
+        )
+        meta = {
+            "crawling_logs": {
+                "page_type": "jobPosting"
+                if self.args.crawl_strategy == JobPostingCrawlStrategy.direct_item
+                else "jobPostingNavigation"
+            },
+        }
+        return Request(
+            url=url,
+            callback=callback,
+            meta=meta,
+        )
+
+    def start_requests(self) -> Iterable[Request]:
+        for url in self.start_urls:
+            yield self.get_start_request(url)
+
+    def parse_navigation(
+        self, response: DummyResponse, navigation: JobPostingNavigation
+    ) -> Iterable[Request]:
+        job_postings = navigation.items or []
+        for request in job_postings:
+            yield self.get_parse_job_posting_request(request)
+
+        if navigation.nextPage:
+            if not job_postings:
+                self.logger.info(
+                    f"Ignoring nextPage link {navigation.nextPage} since there "
+                    f"are no job posting links found in {navigation.url}"
+                )
+            else:
+                yield self.get_nextpage_request(navigation.nextPage)
+
+    def parse_job_posting(
+        self, response: DummyResponse, job_posting: JobPosting
+    ) -> Iterable[JobPosting]:
+        probability = job_posting.get_probability()
+
+        # TODO: convert to a configurable parameter later on after the launch
```

Comment:
We have a middleware in zyte-common-items to drop items below certain probability - what's the advantage of doing it in the spider?

Reply:
Hmm, interesting - so, we're dropping custom attributes if the product probability is low; the middleware in zyte-common-items won't do it.

Reply:
Yeah, I admit I just copied the e-commerce spider and trimmed it down.

We may want to do something differently but also we may want to extract common code, and I think we need to wait until this and articles exist in the same tree.